rollingpin/frontends.py

```diff
@@ -223,9 +235,207 @@ def raw_input(self, prompt):
+
+    def __init__(self, console_input):
+        self.console_input = console_input
+        self.enqueued_pools = set()
+
+    def is_complete(self, hosts):
+        all_pools = {host.pool for host in hosts.iterkeys()}
+        pools_deployed_to = {host.pool for host, status in hosts.iteritems()
+                             if status["status"] in ("complete", "deploying")}
+        untouched_pools = all_pools - pools_deployed_to
+        return not untouched_pools
+
+    @inlineCallbacks
+    def get_next_strategy(self, hosts):
+        pools_deployed_to = {host.pool for host, status in hosts.iteritems()
+                             if status["status"] == "complete"}
+        if len(pools_deployed_to) > 1:
+            print ("Canary deploy complete, please verify the canary "
+                   "hosts are not reporting errors in the logs.")
+            yield prompt_declaration(self.console_input, "The canaries are healthy")
+        else:
+            print ("Canary deploy complete, please verify the canary "
+                   "host is not reporting errors in the logs.")
+            yield prompt_declaration(self.console_input, "The canary is healthy")
+
+        print
+        print "Great!"
+
+        next_strategy = yield get_next_regular_strategy(self.console_input, hosts)
+        returnValue(next_strategy)
+
+
+def round_to_next_target(hosts, raw_target):
+    """Round up from a target percentage to the next achievable value.
+
+    When the number of hosts is low, there's a minimum jump in percentage
+    from host-to-host. This rounds up to the next value so that the user isn't
+    surprised by the deploy going further than expected from the prompts.
+
+    """
+    minimum_step_size = 1. / len(hosts) * 100
+    return int(math.ceil(raw_target / minimum_step_size) * minimum_step_size)
+
+
+@inlineCallbacks
+def get_next_regular_strategy(console_input, hosts):
+    options = [
+        "something's wrong, [a]bort the deploy!",
+    ]
+    strategies = {}
+
+    if len(hosts) <= 3:
+        options.append("roll out to the [n]ext host")
+        strategies["n"] = SingleHostDeployStrategy(console_input)
+    else:
+        percent_complete = calculate_percent_complete(hosts)
+
+        if percent_complete < 65:
+            target = round_to_next_target(hosts, percent_complete + 25)
+            options.append("i'm not worried about load, [j]ump forward to %d%% of hosts" % target)
+            strategies["j"] = PercentDeployStrategy(console_input, target)
+
+        if percent_complete < 85:
+            target = round_to_next_target(hosts, percent_complete + 10)
+            options.append("let's keep an eye on load, [s]tep forward to %d%% of hosts" % target)
+            strategies["s"] = PercentDeployStrategy(console_input, target)
+
+        if percent_complete >= 65:
+            options.append("all is good, [f]inish the deploy")
+            strategies["f"] = FreeDeployStrategy()
```

Comment:
Obligatory observation of inline magical constants.

Reply:
I totally agree on usually avoiding numbers in code, but these are percent thresholds and feel kinda natural as-is plus don't have any scope beyond this function. Do you think giving them symbolic names would be worth the indirection?

Reply:
Dude, you're harshing my nitpick. ;) 

Fine by me as is.